src/module.c

```diff
@@ -13930,6 +13930,75 @@ RedisModuleString *RM_DefragRedisModuleString(RedisModuleDefragCtx *ctx, RedisMo
     return activeDefragStringOb(str);
 }
 
+/* Defrag callback for radix tree iterator, called for each node,
+ * used in order to defrag the nodes allocations. */
+int moduleDefragRaxNode(raxNode **noderef) {
+    raxNode *newnode = activeDefragAlloc(*noderef);
+    if (newnode) {
+        *noderef = newnode;
+        return 1;
+    }
+    return 0;
+}
+
+/* Defragment a Redis Module Dictionary by scanning its contents and calling a value
+ * callback for each value.
+ *
+ * The callback gets the current value in the dict, and should return non-NULL with a new pointer,
+ * if the value was re-allocated to a different address. The callback also gets the key name just as a reference.
+ *
+ * The API can work incrementally by accepting a seek position to continue from, and
+ * returning the next position to seek to on the next call (or return NULL when the iteration is completed).
+ *
+ * This API returns a new dict if it was re-allocated to a new address (will only
+ * be attempted when *seekTo is NULL on entry).
+ */
+RedisModuleDict *RM_DefragRedisModuleDict(RedisModuleDefragCtx *ctx, RedisModuleDict *dict, RedisModuleDefragDictValueCallback valueCB, RedisModuleString **seekTo) {
+    RedisModuleDict *newdict = NULL;
+    raxIterator ri;
+
+    if (*seekTo == NULL) {
+        /* if last seek is NULL, we start new iteration */
+        rax* newrax = NULL;
+        if ((newdict = activeDefragAlloc(dict)))
+            dict = newdict;
+        if ((newrax = activeDefragAlloc(dict->rax)))
+            dict->rax = newrax;
+    }
+
+    raxStart(&ri,dict->rax);
+    if (*seekTo == NULL) {
+        /* assign the iterator node callback before the seek, so that the
+         * initial nodes that are processed till the first item are covered */
+        ri.node_cb = moduleDefragRaxNode;
+        raxSeek(&ri,"^",NULL,0);
+    } else {
+        /* if cursor is non-zero, we seek to the static 'last' */
+        if (!raxSeek(&ri,">", (*seekTo)->ptr, sdslen((*seekTo)->ptr))) {
+            goto cleanup;
+        }
+        /* assign the iterator node callback after the seek, so that the
+         * initial nodes that are processed till now aren't covered */
+        ri.node_cb = moduleDefragRaxNode;
+    }
+
+    while (raxNext(&ri)) {
+        void *newdata = valueCB(ctx, ri.data, ri.key, ri.key_len);
+        if (newdata)
+            raxSetData(ri.node, ri.data=newdata);
+        if (RM_DefragShouldStop(ctx)) {
```

Comment:
```suggestion
        if (++iterator++ > 128 && RM_DefragShouldStop(ctx)) {
```

@oranagra @ephraimfeldblum should we call `RM_DefragShouldStop()` every N times? i think it's reasonable, reported by @ShooterIT

Reply:
yes. certainly. don't want to excessively call the ustime function.

Reply:
Unfortunately, I don't think that will work for me. On the TS side we have dicts of dicts. If a leaf dict should stop, then we cannot afford to let its branch dict continue beyond that point.

Reply:
so maybe in addition to that, we need to raise some flag (return value) to cause the recursion to exit?
or maybe we should aim for users to call these method often, but add some mechanism to throttle the rate at which it calls mstime, and at the same time, cache an exit flag to make it sticky.

maybe it can look at increments of `stat_active_defrag_misses` and `stat_active_defrag_hits`, and check the time once in some 512 pointers?

Reply:
@oranagra Do you mean that it is possible to get to a leaf and if it returns to stop, we just skip all the leaves behind and stop this entire rax defragmentation?
@ephraimfeldblum Is this what you want? What I understand is that your leaf is also a dict, and you don't want to quit in the middle of defragging this leaf.

Reply:
no, i mean something like this:
```c
int RM_DefragShouldStop(RedisModuleDefragCtx *ctx) {
    if (ctx->stopping)
        return 1;
    if (!ctx->endtime)
        return 0;
    if (server.stat_active_defrag_hits - ctx->last_stop_check_hits < 512 &&
        server.stat_active_defrag_misses - ctx->last_stop_check_misses < 1024) {
        return 0;
    }
    ctx->last_stop_check_hits = server.stat_active_defrag_hits;
    ctx->last_stop_check_misses = server.stat_active_defrag_misses;
    if (ctx->endtime < ustime())
        ctx->stopping = 1;
    return ctx->stopping;
}
```

Reply:
@sundb I can pause in the middle of defragging a leaf (though it's unlikely, since the leaves are sets). But if I do, then the parent dict cannot be allowed to continue past that point, it must also stop immediately.
@oranagra's solution looks good to me.